Use Blaze 3.8 (exclusively) (again), and test architectures

moxcodes opened the pull request:
## Proposed changes

Restores the previous commits for implementing the Blaze upgrades, remove the single vectorization flag that appeared to cause the bug in the previous attempt, and tests a selection of architectures for vectorization compilation and test.

The selection of architecture tests is intended to concentrate on those that seemed most likely to get into trouble and to minimize impact on CI runtime (each arch test appears to take around 6 minutes), but I'm happy to modify the selection of compiler/architecture combinations according to suggestions.

Note there is one commit here that will be dropped (it starts with drop:) that switches the CI to use a container on my dockerhub so that CI can run before we update the main container.

(to give appropriate credit: the arch-based vectorization test commit is based on a test written by @nilsdeppe on another branch)

## Upgrade Instructions

<!-- UPGRADE INSTRUCTIONS -->
Upgrade the [Blaze](https://bitbucket.org/blaze-lib/blaze) library to version 3.8.
<!-- UPGRADE INSTRUCTIONS -->

### Types of changes:

- [x] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [x] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

moxcodes committed:
Use Blaze 3.8 (exclusively), remove PointerVector

moxcodes committed:
Update wheeler environment for Blaze 3.8

moxcodes committed:
Add arch-based vectorization tests

moxcodes committed:
Break vector asserts into separate files

moxcodes committed:
fix clang-tidy complaints for VectorImplTestHelper

moxcodes committed:
Disable a vectorization flag to avoid Blaze bug

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+                   -snb,
+                   -hsw,
+                   -skl,
+                   -icx]
```
You can use standard list syntax:

```yaml
sde_arch:
  - "-nhm"
  - "-snb"
  # ...
```

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+          - sde_arch: -icx
+            compiler: gcc-7
+          - sde_arch: -icx
+            compiler: clang-6.0
```
I think you don't need these `exclude`s, the compilers are not in the list above.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
           ctest -j2 --output-on-failure -L inputfiles --repeat after-timeout:3
+
+  arch_datastructures_tests:
+    name: DataStructures arch tests
```
You can omit "tests", it's in this file's name. Perhaps just name this "Archs"

nilsvu commented on tests/Unit/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ add_subdirectory(Options)
 add_subdirectory(Parallel)
 add_subdirectory(ParallelAlgorithms)
 add_subdirectory(PointwiseFunctions)
+add_subdirectory(TestArchitectureVectorization)
```
You can remove the `Test` prefix from this directory name

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+        # - `-Werror`: Treat warnings as error.
+        # - `-march=x86-64`: Make sure we are building on a consistent
+        #   architecture so caching works. This is necessary because GitHub
+        #   may run the job on different hardware.
```
I suggest you remove these comments and refer to the "unit test" job for details instead.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+          -D CHARM_ROOT=/work/charm_6_10_2/multicore-linux-x86_64-${CHARM_CC}
+          -D CMAKE_BUILD_TYPE=Debug
+          -D DEBUG_SYMBOLS=OFF
+          -D RUN_TESTS_IN_TEST_EXECUTABLES=OFF
```
you don't need this line

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+          -D RUN_TESTS_IN_TEST_EXECUTABLES=OFF
+          -D STRIP_SYMBOLS=ON
+          -D STUB_EXECUTABLE_OBJECT_FILES=ON
+          -D STUB_LIBRARY_OBJECT_FILES=ON
```
You probably also don't need this stub functionality, because you only compiling a few files

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+          -D USE_PCH=ON
+          -D USE_CCACHE=ON
+          --warn-uninitialized
+          $GITHUB_WORKSPACE 2>&1 | tee CMakeOutput.txt 2>&1
```
you can remove `--warn-uninitialized` and everything after the `$GITHUB_WORKSPACE` that prints to a file

nilsvu replied:
ping on deleting <s>`2>&1 | tee CMakeOutput.txt 2>&1`</s>

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+        working-directory: /work/build
+        run: |
+          make -j2 TestArchitectureVectorization
+      - name: Print size of build directory
```
you can delete this step I think

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -120,6 +120,7 @@ jobs:
           -D USE_CCACHE=OFF
           -D DEBUG_SYMBOLS=OFF
           -D BUILD_PYTHON_BINDINGS=ON
+          -D BUILD_SHARED_LIBS=OFF
```
did you have to add this line for a particular reason?

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+      fail-fast: false
+      matrix:
+        compiler: [gcc-9,
+                   clang-10
```
For my understanding, how relevant is the choice of compiler for this whole archs-issue?

moxcodes replied:
It is at least somewhat important, because different compilers will bring in different vectorization flags, and that can influence which compile-time branches Blaze uses in the implementation.
I've chosen a couple of the most modern ones under the assumption that these are the most rapidly changing ones, and most prone to make unexpected alterations in how Blaze performs, but I'm happy to make adjustments for which compilers we test on.

nilsvu replied:
Do we have specific cluster/compiler/arch combinations in production that would be important to test, e.g. the case that prompted the revert?

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +434,129 @@ jobs:
+        with:
+          path: /work/ccache
+          key:
+            "ccache-${{ matrix.compiler }}-Debug-pch-ON"
```
Using ccache for these arch tests might not be worth the added complexity in this file. I assume compiling these tests is fast? Then I suggest you remove all ccache-related steps and variables from this job.

nilsvu replied:
Also, we have limited cache space on GitHub, which we should probably preferably use for the unit tests.

nilsvu reviewed (changes requested):
I only reviewed the commit "Add arch-based vectorization tests"

kidder commented on docs/DevGuide/ImplementingVectors.md:
```diff
@@ -45,14 +45,26 @@ DataVector>`. This template pattern is known as the
 may be inherited calling the following alias code in the vector class
-definition:
-
+ definition:
```
remove the leading space

kidder commented on docs/DevGuide/ImplementingVectors.md:
```diff
@@ -64,7 +64,7 @@ Blaze types, but in addition, the methods `set_data_ref`, and `pup` are defined
 will be implicitly inherited from `VectorImpl`. The assignment and constructors
 may be inherited calling the following alias code in the vector class
- definition:
+definition:
```
most of this commit belongs somewhere else...

kidder commented on tests/Unit/TestArchitectureVectorization/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+#
+# It can be used to test that vectorization with Blaze correctly works on all
+# of our target architectures without needing access to the CPUs. This is
+# further made possible by the Intel Software Development Emulator.
```
omit 'further'

kidder commented on tests/Unit/TestArchitectureVectorization/TestArchitectureVectorization.cpp:
```diff
@@ -0,0 +1,53 @@
+#include <string>
+
+#include "ErrorHandling/FloatingPointExceptions.hpp"
+#include "Framework/SetupLocalPythonEnvironment.hpp"
```
is this needed?

kidder commented on tests/Unit/TestArchitectureVectorization/TestArchitectureVectorization.cpp:
```diff
@@ -0,0 +1,53 @@
+  // that we only initialize and finalize the python env once. Initialization is
+  // done in the constructor of SetupLocalPythonEnvironment, while finalization
+  // is done in the constructor of RunTests.
+  pypp::SetupLocalPythonEnvironment::finalize_env();
```
is pypp used in the above tests?

kidder commented on tests/Unit/TestArchitectureVectorization/TestArchitectureVectorization.cpp:
```diff
@@ -0,0 +1,53 @@
+
+#include "tests/Unit/RunTests.def.h"  /// IWYU pragma: keep
+
+// Needed for tests that use the GlobalCache since it registers itself with
```
is this needed?

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -436,19 +435,20 @@ jobs:
-                   -skl,
-                   -icx]
+        sde_arch:
+          - "-nhm"
```
(nitpicking) The list of compilers above can also be formatted like this

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -463,43 +463,19 @@ jobs:
+        # See the unit test job for the reasons for these configuration choices
         CXXFLAGS: "-Werror"
-        # We make sure to use a fixed absolute path for the ccache directory
         CCACHE_DIR: /work/ccache
```
delete these (now unused) ccache vars

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -531,13 +507,10 @@ jobs:
           -D CHARM_ROOT=/work/charm_6_10_2/multicore-linux-x86_64-${CHARM_CC}
           -D CMAKE_BUILD_TYPE=Debug
```
Can arch-related issues occur in `Release` mode but not in `Debug`?

moxcodes replied:
I suppose it might be possible in principle, but I don't believe that Blaze does any of its vectorization any differently for debug builds.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -531,13 +507,10 @@ jobs:
           -D DEBUG_SYMBOLS=OFF
-          -D RUN_TESTS_IN_TEST_EXECUTABLES=OFF
           -D STRIP_SYMBOLS=ON
           -D STUB_EXECUTABLE_OBJECT_FILES=ON
```
Also delete these lines (I think they just save disk space, but you're compiling only a few files)

nilsdeppe replied:
So why remove them?

nilsvu replied:
🤷‍♂️ seems overkill for those few compiled libs.. but can't hurt I guess

nilsvu commented on tests/Unit/ArchitectureVectorization/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+  ../DataStructures/Test_ComplexDataVectorBinaryOperations.cpp
+  )
+
+# Modify which libraries to link for your test case.
```
delete this comment

nilsvu commented on tests/Unit/ArchitectureVectorization/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+  DataStructures
+  )
+
+# Core required libraries
```
delete this comment

nilsdeppe commented:
What's generally not clear to me is that having a new build target for each individual arch in CI is the right approach. A couple of reasons:
1. we have to pull the container, etc. more often (takes extra human time to wait)
2. it's basically impossible to run the arch tests locally

What do you think of having a bash script that takes as arguments the source dir, target dir where to create all the build dirs and then loops over all the archs to test them?

moxcodes commented:
Posted an additional fixup for the new comments, and condensing the tests to just make one run per compiler to save on the time for pulling the container. That looks like it takes the time per arch from around 6m to around 4m on average, so I added two more compilers as well.
Cheers!

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -440,88 +440,89 @@ jobs:
+          for ARCH_PARAM in ${ARCH_PARAM_LIST[@]}; do
+            OVERRIDE_ARCH=`echo ${ARCH_PARAM} | cut -d";" -f2`
+            SDE_FLAG=`echo ${ARCH_PARAM} | cut -d";" -f1`
+            echo $OVERRIDE_ARCH
```
I think these diagnostic echos would be good to keep (I assume they're relics from debugging). What if you did `"CMake arch flag: $OVERRIDE_ARCH"` and `"Intel SDE arch flag: $SDE_FLAG"`? Since we will inevitably have to maintain this list, it would probably make life easier for people if some info is printed :)

moxcodes replied:
Yeah, this is a pretty quick way to understand which of the archs failed, so I thought it would be good to keep. I've added the extra information you suggested.

kidder commented:
you have my approval to squash when @nilsdeppe and @nilsleiffischer agree

moxcodes commented:
(squashed in one more minor change to the github CI file that I had intended -- actually do the Arch builds on only 4 of the target compilers, rather than all of them)

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -440,88 +440,80 @@ jobs:
+            -D CMAKE_BUILD_TYPE=Debug\
+            -D DEBUG_SYMBOLS=OFF\
+            -D STRIP_SYMBOLS=ON\
+            -D STUB_EXECUTABLE_OBJECT_FILES=ON\
```
Either use both flags `STUB_EXECUTABLE_OBJECT_FILES` and `STUB_LIBRARY_OBJECT_FILES` or neither. In this comment https://github.com/sxs-collaboration/spectre/pull/2574#discussion_r515536029 I suggested to remove them because they probably don't save any significant disk space for the few files this test compiles, but I guess it also can't hurt to keep them. Just be consistent.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -440,88 +440,80 @@ jobs:
+            make -j2 TestArchitectureVectorization
+
+            /work/sde/sde ${SDE_FLAG}\
+            -- ./bin/TestArchitectureVectorization [Unit]
```
Will the test fail if this command fails for a particular arch? Please make sure.

moxcodes replied:
Yes, I've checked that this test gets the red x on the github actions when the vectorization fix from this PR is removed.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +433,87 @@ jobs:
           # it is a random timeout or systematic
           ctest -j2 --output-on-failure -L inputfiles --repeat after-timeout:3
+
+  arch_datastructures_tests:
```
(picking a line at random) Does the Blaze upgrade resolve https://github.com/sxs-collaboration/spectre/issues/1807? (we can also check this once this PR is done)

moxcodes replied:
I'm not sure at what point the issue was resolved, but I've dropped the offending vectorization flag from the debug builds in one of the fixups, and the CI seems to run fine.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +433,87 @@ jobs:
+          - "gcc-7"
+          - "gcc-9"
+          - "clang-6.0"
+          - "clang-10"
```
Do you think this is the smallest possible number of CI builds, or could we reduce the number of compiler configurations further to speed up CI? Related to this comment https://github.com/sxs-collaboration/spectre/pull/2574/#discussion_r515477632:
> Do we have specific cluster/compiler/arch combinations in production that would be important to test, e.g. the case that prompted the revert?

moxcodes replied:
The answer here is that it is not terribly clear. The single case we have encountered is in the most modern compiler, with one of the latest architectures, so that tends to indicate that 'new is buggy', but I could also see an argument that the combination of oldest compiler with newest vectorization flags could be concerning. So, I've picked the oldest and newest supported compilers for each of clang and gcc, but the pick is pretty arbitrary. It at least certainly covers the single case that failed previously.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +433,87 @@ jobs:
+      fail-fast: false
+      matrix:
+        compiler:
+          - "gcc-7"
```
(nitpicking, not requesting this change) The quotes `"` are not needed in this list (I included them in my original suggestion for the arch list because those started with a dash `-`, so I wasn't sure if yaml could handle the syntax)

nilsvu commented on src/Utilities/Blaze.hpp:
```diff
@@ -41,4 +38,75 @@ constexpr bool useDefaultInitialization = false;
+
+namespace blaze {
+template <typename T>
+BLAZE_ALWAYS_INLINE SIMDdouble step_function(const SIMDf64<T>& v) noexcept
```
Could you add a few comments why these additions are necessary?

moxcodes replied:
It's less of an addition, and more just copying over the last piece of PointerVector.hpp that we can't immediately discard, but I can add a code comment explaining why we need it.

nilsvu commented on tests/Unit/DataStructures/Test_DenseMatrixAsserts.cpp:
```diff
@@ -0,0 +1,15 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
```
Could you add a comment here like:

```cpp
/// \file
/// These error tests are in a separate file so `Test_DenseMatrix.cpp` can be
/// compiled into `TestArchitectureVectorization`
```

(similarly for the other files)

nilsvu replied:
ping: this file is missing the note

nilsvu commented on cmake/SetupCxxFlags.cmake:
```diff
@@ -24,12 +24,12 @@ if(NOT "${OVERRIDE_ARCH}" STREQUAL "OFF")
       APPEND PROPERTY
       INTERFACE_COMPILE_OPTIONS
-      $<$<COMPILE_LANGUAGE:CXX>:-march=${OVERRIDE_ARCH}>)
+      $<$<COMPILE_LANGUAGE:CXX>:-march=${OVERRIDE_ARCH} -mno-avx512f>)
```
Could you add a comment why this flag is here

nilsvu commented on src/Utilities/Blaze.hpp:
```diff
@@ -18,15 +21,9 @@
-}
+const blaze::AlignmentFlag blaze_unaligned = blaze::AlignmentFlag::unaligned;
+const blaze::PaddingFlag blaze_unpadded = blaze::PaddingFlag::unpadded;
 
```
Do you think it would make sense to add a type alias for the `blaze::Group<0>` that you use in a few places? Then you could add some comments what the group is used for. Perhaps also add a few words to `ImplementingVectors.md`

nilsvu commented on src/Utilities/Blaze.hpp:
```diff
@@ -18,15 +21,9 @@
-constexpr bool useDefaultInitialization = false;
-}
+const blaze::AlignmentFlag blaze_unaligned = blaze::AlignmentFlag::unaligned;
+const blaze::PaddingFlag blaze_unpadded = blaze::PaddingFlag::unpadded;
```
Not sure these two variables are necessary. I find `blaze::AlignmentFlag::unaligned` as clear as `blaze_unaligned` and not too much longer, and you avoid the level of indirection.

Edit: I now realize that these variables were previously defined by blaze and now replaced by the enum? Still, using the enum directly seems more clear to me.

nilsvu reviewed (changes requested):
My comments are mostly requests for more comments and documentation, so we remember the reasons for these design decisions

nilsdeppe reviewed (commented):
I think this is a much better design. What's still not clear (and can be discussed without code change) what the advantage of having the for loops, build dir creation, etc. in the yaml file is over having a bash script in `tools/`. In the latter case it would be easier to test locally and to switch CI systems. It might also make it somewhat easier to maintain a list inside the `tools/` script than inside a file in a hidden directory :) Of course, there might be a good reason I've overlooked for not doing that!

moxcodes commented:
@nilsdeppe In this case, I mostly anticipated that the precise format that I use to get the github 'matrix' flags into the bash script are probably pretty different from what I'd use in a dedicated bash script. I suppose I could parse those out and then call a tools/ bash script inside the loop, but then it seems kindof silly to just put the cmake and build commands as a whole extra file.
My feeling is that if we want to have a local test that does everything we want, that's not an unreasonable tool to have, but I'd script it a bit differently to allow the user to specify more human-accessible command line options.

moxcodes commented:
new fixups posted (all new ones start with `fixup 2:`).
Cheers!

nilsdeppe commented on tests/Unit/ArchitectureVectorization/CMakeLists.txt:
```diff
@@ -0,0 +1,66 @@
+target_link_libraries(
+  ${EXECUTABLE}
+  PRIVATE
+  DataStructures
```
[optional, and my fault] you can combine the `target_link_libraries` if you wish. In `RunSingleTest` they're separated so that the first section is the "User needs to edit" part and the second section is "We need this in order to do anything"

nilsdeppe reviewed (commented):
This all looks good to me. Once you've squashed and rebased I'll test on my machine and push the new container

nilsvu reviewed (commented):
Please squash, including this minor comment

moxcodes commented:
Everything has been squashed down. I've also dropped the commit that made github actions use my personal dockerhub container, so the CI will now fail until the sxs container is updated.
Cheers!

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +429,88 @@ jobs:
+              "-skl;skylake" "-icx;icelake-server" "-tgl;tigerlake")
+            compiler: clang-10
+    container:
+      image: moxcodes/spectrebuildenv:archtests
```
This test still uses your container

moxcodes replied:
aw, shoot -- the trick of just dropping that commit didn't work with the squashes. fixing now...

nilsdeppe reviewed (commented):
This builds on my machine, but the commits are not ordered correctly. The commit `Disable a vectorization flag to avoid Blaze bug` needs to be first otherwise operations like git bisect will become useless for possibly months to come because not each individual commit builds.

moxcodes commented:
Commits reordered.
Cheers!

nilsvu commented on src/DataStructures/Variables.hpp:
```diff
@@ -350,13 +351,13 @@ class Variables<tmpl::list<Tags...>> {
   friend SPECTRE_ALWAYS_INLINE decltype(auto) operator-(
-      const blaze::Vector<VT, VF>& lhs, const Variables& rhs) noexcept {
-    return ~lhs - rhs.variable_data_;
+      const blaze::DenseVector<VT, VF>& lhs, const Variables& rhs) noexcept {
```
Why do you specialize these arguments from `Vector` to `DenseVector`? Shouldn't these operations work fine for any Blaze vector type?

I'm asking because Blaze features such as [`subvector`](https://bitbucket.org/blaze-lib/blaze/wiki/Subvectors) return views into vectors represented as some other vector type in the inheritance hierarchy (I think `DynamicVector`). Specialising these arguments too much would break compatibility with that.

moxcodes replied:
My understanding is that these overloads are present primarily to allow operations between `Variables` to work with the expression templates. It does have the upshot of allowing operations between raw blaze types and `Variables`, but I don't think we'd want to encourage that use. If we did want to support vector x Variables operations, we'd want to do so through our own vector types.

Practically, I made this change because it was the most direct way to restore functionality for the expression templates when I changed the underlying vector type to `CustomVector` instead of `PointerVector` -- the expression templates we get from the `CustomVector` would require (at least) a bunch of type-traits like those in `VectorImpl`. That might be the extent of additional changes, but I haven't delved deeply in that direction to work out whether more annoying blaze compilation errors crop up if those are fixed.

nilsdeppe replied:
I believe the change @moxcodes made is correct

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +429,88 @@ jobs:
+              fi
+            fi
+
+            cmake\
```
you don't need the `\` line-continuations since the `>` in line 470 means that non-empty lines are concatenated

moxcodes replied:
This is incorrect for this particular scenario. I think it must be the for loop that prevents the usual automatic line continuation.

nilsvu replied:
hm, then I don't exactly understand how the bash and yaml syntax interact here, but if this works let's leave it as-is 🤷‍♂️

nilsdeppe commented:
@kidder @nilsleiffischer could you test this on your MacBooks to make sure everything works there?

nilsdeppe commented:
@kidder @nilsleiffischer could you test this on your MacBooks to make sure everything works there?

nilsvu commented:
I can confirm it works on macOS 11.0.1 with AppleClang 12.0.0 and with Blaze 3.8 installed via Spack.

nilsdeppe commented:
Thanks for checking!

@kidder could you also test on your machine?

nilsvu commented:
@moxcodes you may have already seen this: Blaze 3.7 introduces [a `solve` function](https://bitbucket.org/blaze-lib/blaze/wiki/Matrix%20Operations#!linear-systems) that calls into LAPACK's `gesv` and friends, so if you like you can use it to replace your LAPACK wrappers in `NumericalAlgorithms/LinearSolver/Lapack.hpp`.

moxcodes commented:
@nilsleiffischer Thanks for pointing that out! I think I probably can switch that out -- I'll do that in a separate PR to avoid cluttering this one too much. There may be other places we currently use LAPACK routines explicitly where we could use blaze replacements.

@nilsdeppe @kidder Let me know when you think it's time to move forward with this PR and I'll rebase and fix the conflicts.
Cheers!

kidder commented:
I think it is time to move forward; @nilsdeppe ?

nilsdeppe commented:
Yes

moxcodes committed:
fix NewtonianEuler test with unset parameter

moxcodes commented:
Sorry for the delay -- I've rebased this and solved the conflicts. It, of course, fails CI because of the needed container changes. For some confidence, this branch: https://github.com/moxcodes/spectre/actions/runs/481972447 succeeds, and differs only by the container switch.

nilsdeppe reviewed (dismissed):
LGTM.

@nilsleiffischer @kidder could you please take a look? Once we're happy with the PR I'll update the container

nilsvu commented on src/DataStructures/Variables.hpp:
```diff
@@ -107,8 +107,9 @@ class Variables<tmpl::list<Tags...>> {
   using pointer_type =
-      PointerVector<value_type, blaze_unaligned, blaze_unpadded, transpose_flag,
-                    blaze::DynamicVector<value_type, transpose_flag>>;
+      blaze::CustomVector<value_type, blaze::AlignmentFlag::unaligned,
```
This file is missing an include for `CustomVector`, `DenseVector` and other Blaze types. Please add if you get the chance, but I don't want this to hold up the PR.

nilsvu commented on src/Utilities/Blaze.hpp:
```diff
@@ -41,4 +37,77 @@ constexpr bool useDefaultInitialization = false;
+
+#define BLAZE_USE_STREAMING 1
+#define BLAZE_USE_OPTIMIZED_KERNELS 1
+#define BLAZE_USE_DEFAULT_INITIALIZATON 0
```
Question: Is this the right place to define these macros, or should we add them to the `Blaze` imported target in `SetupBlaze` in the build system (with [target_compile_definitions](https://cmake.org/cmake/help/latest/command/target_compile_definitions.html)? Can the include order mess things up here?

moxcodes replied:
So, I think the suggestion of using compile flags is the 'right' way to use these blaze flags, but our choices in cmake make this difficult for us. In particular, any compile definitions that are set outside the pre-compiled header must similarly be set for every compilation that used the pre-compiled header. I was unable to find a clear way to propagate those compile definitions in a straightforward way that did not result in a `-winvalid-pch` from gcc when using a pre-compiled header.
I'm open to suggestions, but because this is how we've handled blaze up untill now, my inclination is to enforce the include order consistently in our source files and include a code comment where applicable, but leave the cmake best practices aside for now, particularly because (based on our pch handling) the cmake pre-compiled header support seems somewhat incomplete.

nilsvu replied:
I think the PCH lib just needs to link `PUBLIC`ly with `Blaze` so it propagates the flags. This seems to work on develop in the container with gcc-9:

```diff
diff --git a/cmake/SetupBlaze.cmake b/cmake/SetupBlaze.cmake
index d35a636fc..6f1396805 100644
--- a/cmake/SetupBlaze.cmake
+++ b/cmake/SetupBlaze.cmake
@@ -17,6 +17,22 @@ set_property(TARGET Blaze PROPERTY
 set_property(TARGET Blaze PROPERTY
   INTERFACE_LINK_LIBRARIES Lapack)
 
+target_compile_definitions(
+  Blaze
+  INTERFACE
+  # Override SMP configurations
+  _BLAZE_SYSTEM_SMP_H_
+  BLAZE_USE_SHARED_MEMORY_PARALLELIZATION=0
+  BLAZE_OPENMP_PARALLEL_MODE=0
+  BLAZE_CPP_THREADS_PARALLEL_MODE=0
+  BLAZE_BOOST_THREADS_PARALLEL_MODE=0
+  # Disable MPI parallelization
+  _BLAZE_SYSTEM_MPI_H_
+  BLAZE_MPI_PARALLEL_MODE=0
+  # Disable HPX parallelization
+  BLAZE_HPX_PARALLEL_MODE=0
+  )
+
 add_interface_lib_headers(
   TARGET Blaze
   HEADERS
diff --git a/cmake/SetupPch.cmake b/cmake/SetupPch.cmake
index 5313d6a3a..c9f4fe273 100644
--- a/cmake/SetupPch.cmake
+++ b/cmake/SetupPch.cmake
@@ -177,7 +177,7 @@ if (USE_PCH)
     )
   target_link_libraries(
     ${SPECTRE_PCH_LIB}
-    PRIVATE
+    PUBLIC
     Blaze
     Brigand
     )
diff --git a/src/Utilities/Blaze.hpp b/src/Utilities/Blaze.hpp
index 5d8b11bfb..02dd3ef87 100644
--- a/src/Utilities/Blaze.hpp
+++ b/src/Utilities/Blaze.hpp
@@ -27,18 +27,4 @@ constexpr bool useOptimizedKernels = true;
 namespace blaze {
 constexpr bool useDefaultInitialization = false;
 }
-
-// Override SMP configurations
-#define _BLAZE_SYSTEM_SMP_H_
-#define BLAZE_USE_SHARED_MEMORY_PARALLELIZATION 0
-#define BLAZE_OPENMP_PARALLEL_MODE 0
-#define BLAZE_CPP_THREADS_PARALLEL_MODE 0
-#define BLAZE_BOOST_THREADS_PARALLEL_MODE 0
-
-// Disable MPI parallelization
-#define _BLAZE_SYSTEM_MPI_H_
-#define BLAZE_MPI_PARALLEL_MODE 0
-
-// Disable HPX parallelization
-#define BLAZE_HPX_PARALLEL_MODE 0
 /// \endcond
```

moxcodes replied:
On a bare build directory, the patch you posted generates gcc warnings:
```
cc1plus: warning: /home/mox/research/SXS/spectre/spectre-build-fresh/SpectrePch.hpp.gch: not used because `BLAZE_BOOST_THREADS_PARALLEL_MODE' not defined [-Winvalid-pch]
```
my cmake call was:
```
cmake -D CHARM_ROOT=/work/charm_6_10_2/multicore-linux-x86_64-gcc/ -D CMAKE_CXX_COMPILER=/usr/bin/g++-9 -D CMAKE_C_COMPILER=/usr/bin/gcc-9 -D CMAKE_Fortran_COMPILER=/usr/bin/gfortran-7 -D USE_PCH=ON /home/mox/research/SXS/spectre/
```
(critically, I think it is necessary to force a rebuild of your PCH to see the behavior -- I needed to clear my build directory)

nilsvu commented on tests/Unit/DataStructures/Test_VectorImpl.cpp:
```diff
@@ -37,24 +37,3 @@ static_assert(not is_derived_of_vector_impl_v<std::complex<double>>,
-  (void)(ComplexDataVector{} != std::complex<double>{});
-  ERROR("Failed to trigger ASSERT in an assertion test");
-#endif
-}
```
This deletion seems to be in the wrong commit, so please correct if you get the chance, but no reason to hold up the PR.

moxcodes replied:
I believe this tested a piece of functionality that is now delegated to blaze, so no spectre assert is triggered for this error condition, so I think it makes sense in the 'remove PointerVector' commit.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -303,11 +303,6 @@ jobs:
-          # See issue https://github.com/sxs-collaboration/spectre/issues/1807
-          - build_type: Debug
-            EXTRA_CXX_FLAGS: "-DBLAZE_USE_VECTORIZATION=0"
-          # Test with ASAN
```
Don't delete this single line please

nilsvu commented on tests/Unit/ArchitectureVectorization/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+  DataStructures
+  Framework
+  Informer
+  )
```
Add `SystemUtilities`, `ErrorHandling`, `Parallel # for printf`

kidder replied:
alphabetical

nilsdeppe replied:
why are we noting that `Parallel` is for printf here but we never do this anywhere else in the code?

kidder commented:
I have no requested changes and this can be merged without my explicit approval

moxcodes commented:
I've adjusted the `Blaze.hpp` macros after reviewing the blaze library use of the compile definitions, and propagated the include order enforcement in a fixup. I think that compile order doesn't matter for PCH builds, but consistency is important.
Cheers!

moxcodes commented:
rebased following #2778. I believe all of the pending comments are now addressed in fixups.
Cheers!
edit: here's a link to the CI run that contains the same content as the current version, except it uses my updated dockerhub container: https://github.com/moxcodes/spectre/actions/runs/496722885

kidder commented:
@nilsleiffischer please look at the fixups

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -183,6 +183,7 @@ jobs:
           -D CMAKE_BUILD_TYPE=${{ matrix.build_type }}
           -D OVERRIDE_ARCH=x86-64
           -D USE_CCACHE=OFF
+          -D USE_PCH=OFF
```
you should be able to drop this commit now

nilsvu commented on cmake/SetupBlaze.cmake:
```diff
@@ -20,6 +20,21 @@ set_property(TARGET Blaze PROPERTY
+  BLAZE_OPENMP_PARALLEL_MODE=0
+  BLAZE_CPP_THREADS_PARALLEL_MODE=0
+  BLAZE_BOOST_THREADS_PARALLEL_MODE=0
+  BLAZE_USE_SHARED_MEMORY_PARALLELIZATION=0
```
This flag is duplicate, you can delete this line.

Also, it wouldn't hurt to carry over the comments from the `Blaze.hpp` to this cmake file.

nilsvu commented on src/Utilities/Blaze.hpp:
```diff
@@ -4,41 +4,80 @@
 
+#include <blaze/math/CustomVector.h>
+#include <blaze/math/DenseVector.h>
+#include <blaze/math/GroupTag.h>
```
I believe these should be added below the `#pragma once`

nilsvu commented on src/Utilities/Blaze.hpp:
```diff
@@ -4,41 +4,80 @@
+
 #pragma once
 
 #ifdef __GNUC__
 #pragma GCC system_header
 #endif
```
(optional) These lines should probably be deleted, but if that leads to warnings I'm fine with keeping them around for now

nilsvu reviewed (commented):
Looks good, please squash, including these minor things

moxcodes commented:
fixups squashed in. This is the CI run with these changes that uses my dockerhub container: https://github.com/moxcodes/spectre/actions/runs/510539246.
Cheers!

nilsvu reviewed (dismissed):
LGTM. @nilsdeppe and @kidder could you take over, push the container and merge when you're ready? No need to get my re-approval if anything changes.

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -17,6 +20,23 @@ set_property(TARGET Blaze PROPERTY
+  BLAZE_USE_PADDING=0
+  BLAZE_USE_STREAMING=1
+  BLAZE_USE_OPTIMIZED_KERNELS=1
+  BLAZE_USE_DEFAULT_INITIALIZATON=0
```
You've added comments for ones where I understand what the flag does quite easily, but not the trickier ones. Specifically, `USE_DEFAULT_INITIALIZATION` isn't documented in Blaze, or at least not where I would expect.

nilsdeppe reviewed (commented):
@kidder should we have @moxcodes rebase and then update the container tonight/tomorrow?

kidder commented:
I'm okay, please rebase and squash things, then we can update the container

moxcodes commented:
rebased, and squashed in @nilsdeppe 's suggestions.
Here's the CI run for the same changes, but uses my container: https://github.com/moxcodes/spectre/actions/runs/532324962
Cheers!

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -131,7 +131,7 @@ RUN wget https://github.com/Kitware/CMake/releases/download/v3.12.4/cmake-3.12.4
 # Build with GCC7 when easily possible to maximize ABI compatibility.
```
[random line in this file] I also needed to change:
```
&& curl https://bootstrap.pypa.io/get-pip.py --output get-pip.py \ 
```
to
```
&& curl https://bootstrap.pypa.io/2.7/get-pip.py --output get-pip.py \ 
```
above. Could you add a commit at the end that does that?

moxcodes committed:
Update pip link for container

moxcodes commented:
@nilsdeppe done.

nilsdeppe reviewed (approved):
Thank you :)
